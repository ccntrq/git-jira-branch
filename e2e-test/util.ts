import {type SpawnSyncReturns, execSync, spawnSync} from 'node:child_process';
import {rmSync} from 'node:fs';
import {mkdtemp} from 'node:fs/promises';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {Brand} from 'effect';

export type Directory = string & Brand.Brand<'Directory'>;
export const Directory = Brand.nominal<Directory>();

export const setupTmpDir = async (): Promise<
  readonly [Directory, () => void]
> => {
  const tmpDir = await mkdtemp(join(tmpdir(), 'git-jira-branch-e2e-'));
  // biome-ignore lint/nursery/noConsole: log statement okay in this test
  console.debug(`Setting up e2e env in ${tmpDir}`);

  execSync('git init', {cwd: tmpDir});

  execSync('git config user.email "lester-tester@example.com"', {
    cwd: tmpDir,
  });
  execSync('git config user.name "Lester Tester"', {cwd: tmpDir});
  execSync('git commit -m "init" --allow-empty', {cwd: tmpDir});

  const cleanup = () => {
    // biome-ignore lint/nursery/noConsole: log statement okay in test
    console.debug(`Cleaning up e2e env in ${tmpDir}`);
    rmSync(tmpDir, {force: true, recursive: true});
  };

  return [Directory(tmpDir), cleanup] as const;
};

export const runApp = (dir: Directory, ...args: Array<string>): string => {
  const cmd = `git-jira-branch ${args.join(' ')}`;
  return execSync(cmd, {cwd: dir}).toString();
};

export const spawnApp = (
  dir: Directory,
  ...args: Array<string>
): SpawnSyncReturns<Buffer> => {
  const cmd = `git-jira-branch ${args.join(' ')}`;
  return spawnSync(cmd, {
    cwd: dir,
    shell: true,
  });
};

export const currentBranch = (dir: Directory): string =>
  execSync('git branch --show-current', {cwd: dir}) //
    .toString()
    .trim();

export const createBranch = (dir: Directory, name: string): string =>
  execSync(`git checkout -b ${name}`, {cwd: dir}) //
    .toString()
    .trim();

export const switchBranch = (dir: Directory, name: string): string =>
  execSync(`git switch ${name}`, {cwd: dir}) //
    .toString()
    .trim();

export const createCommit = (dir: Directory, message: string): string =>
  execSync(`git commit -m "${message}" --allow-empty`, {cwd: dir}) //
    .toString()
    .trim();
