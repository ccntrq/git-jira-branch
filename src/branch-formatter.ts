import type {GitBranch} from './types';

import {Doc} from '@effect/printer';
import {green} from '@effect/printer-ansi/Ansi';
import {render} from '@effect/printer-ansi/AnsiDoc';
import {Chunk, pipe} from 'effect';

export const formatBranches = (branches: Chunk.Chunk<GitBranch>): string => {
  const hasCurrent = Chunk.some(branches, (_) => _.isCurrent);
  return pipe(
    branches,
    Chunk.map((branch) =>
      branch.isCurrent
        ? pipe(Doc.text('*'), Doc.catWithSpace(mkGreen(Doc.text(branch.name))))
        : Doc.indent(Doc.text(branch.name), hasCurrent ? 2 : 0),
    ),
    Doc.vcat,
    render({style: 'pretty'}),
  );
};

const mkGreen = Doc.annotate(green);
