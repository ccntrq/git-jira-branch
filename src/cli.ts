import {ValidationError} from '@effect/cli';
import type {CliApp} from '@effect/cli/CliApp';
import * as Command from '@effect/cli/Command';
import * as HelpDoc from '@effect/cli/HelpDoc';
import * as Span from '@effect/cli/HelpDoc/Span';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Console, Effect, pipe} from 'effect';

import * as packageJson from '../package.json';
import {create} from './commands/create/create.command';
import {deleteCommand} from './commands/delete/delete.command';
import {info} from './commands/info/info.command';
import {list} from './commands/list/list.command';
import {open} from './commands/open/open.command';
import {switchCommand} from './commands/switch/switch.command';
import {tidy} from './commands/tidy/tidy.command';
import type {NoAssociatedBranch} from './schema/no-associated-branch';
import type {AppConfigService} from './services/app-config';
import type {GitClient} from './services/git-client';
import type {JiraClient} from './services/jira-client';
import type {GitJiraBranchError} from './types';

// for version and help
const gitJiraBranch = pipe(Command.make('git-jira-branch', {}));

const mainCommand = gitJiraBranch.pipe(
  Command.withSubcommands([
    create,
    switchCommand,
    deleteCommand,
    open,
    info,
    list,
    tidy,
  ]),
);

const cli = {
  name: packageJson.name,
  version: packageJson.version,
  summary: Span.text(packageJson.description),
};

export const cliEffect = (
  args: Array<string>,
): Effect.Effect<
  void,
  GitJiraBranchError | ValidationError.ValidationError | NoAssociatedBranch,
  | CliApp.Environment
  | GitClient
  | AppConfigService
  | JiraClient
  | CommandExecutor.CommandExecutor
> =>
  Command.run(
    mainCommand,
    cli,
  )(args).pipe(
    Effect.tapError((e) => {
      if (ValidationError.isValidationError(e)) {
        // handled and printed by the cli library already
        return Effect.void;
      }
      return printErrors(HelpDoc.p(Span.error(e.message)));
    }),
  );

const printErrors = (doc: HelpDoc.HelpDoc): Effect.Effect<void> =>
  Console.error(HelpDoc.toAnsiText(doc));
