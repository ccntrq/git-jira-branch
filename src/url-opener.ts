import * as ShellCommand from '@effect/platform/Command';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Effect, pipe} from 'effect';
import {UsageError} from './types';

export const openUrl = (
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  Effect.gen(function* () {
    const platform = yield* getPlatform();

    if (platform === 'linux') {
      return yield* xdgOpen(url);
    }

    // TODO: add support for opening urls on other platforms
    return yield* Effect.fail(
      UsageError({
        message: `Opening urls on ${platform} is not yet supported.`,
      }),
    );
  });

const getPlatform = (): Effect.Effect<NodeJS.Platform> =>
  Effect.succeed(process.platform);

const xdgOpen = (
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  pipe(
    ShellCommand.make('xdg-open', url),
    ShellCommand.exitCode,
    Effect.mapError(() =>
      UsageError({message: 'Browser could not be opened.'}),
    ),
    Effect.map(() => void 0),
  );
