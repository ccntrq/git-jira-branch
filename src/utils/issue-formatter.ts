import type {JiraIssue} from '../types';

import {Doc} from '@effect/printer';
import {bold, underlined} from '@effect/printer-ansi/Ansi';
import {render} from '@effect/printer-ansi/AnsiDoc';
import {pipe} from 'effect';

export const formatIssue = (issue: JiraIssue): string => {
  const heading = pipe(
    [
      mkBold(Doc.text(issue.key)),
      pipe(
        [Doc.char('-'), Doc.reflow(issue.fields.summary)],
        unwords,
        Doc.hang(2),
      ),
    ],
    Doc.seps,
    Doc.hang(2),
    underline,
  );

  const details = pipe(
    [
      mkBold(Doc.text(issue.fields.issuetype.name)),
      unwords([
        Doc.char('|'),
        mkBold(Doc.text('Status:')),
        Doc.text(issue.fields.status.name),
      ]),
      unwords([
        Doc.char('|'),
        mkBold(Doc.text('Creator:')),
        Doc.text(issue.fields.creator.displayName),
      ]),
      issue.fields.assignee
        ? unwords([
            Doc.char('|'),
            mkBold(Doc.text('Assignee:')),
            Doc.text(issue.fields.assignee.displayName),
          ])
        : Doc.empty,
    ],
    Doc.seps,
    Doc.hang(2),
    underline,
  );

  const description = issue.fields.description
    ? pipe(
        issue.fields.description.split(/\n|\r\n/),
        (lines) => lines.map((l) => Doc.reflow(l)),
        Doc.vcat,
      )
    : Doc.empty;

  const doc = pipe(
    [
      heading,
      details,
      ...(Doc.isEmpty(description) ? [] : [Doc.empty, description]),
    ],
    unlines,
  );

  return render(doc, {style: 'pretty', options: {lineWidth: 80}});
};

const unwords = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithSpace)(docs);
const unlines = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithLine)(docs);
const mkBold = Doc.annotate(bold);
const underline = Doc.annotate(underlined);
