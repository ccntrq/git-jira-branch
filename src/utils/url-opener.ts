import * as ShellCommand from '@effect/platform/Command';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Effect, Option, pipe} from 'effect';
import {UsageError} from '../types';

export const openUrl = (
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  Effect.gen(function* () {
    const platform = yield* getPlatform();
    const command = getCommand(platform);

    return yield* Option.match(command, {
      onNone: () =>
        Effect.fail(UsageError({message: `Unsupported platform: ${platform}`})),
      onSome: (command) => openCommand(command, url),
    });
  });

const getPlatform = (): Effect.Effect<NodeJS.Platform> =>
  Effect.succeed(process.platform);

const getCommand = (platform: NodeJS.Platform): Option.Option<string> => {
  const commands: Partial<Record<NodeJS.Platform, string>> = {
    darwin: 'open',
    linux: 'xdg-open',
    // win32: 'start', /* TODO: Implement opening for Windows */
  };
  return Option.fromNullable(commands[platform]);
};

const openCommand = (
  command: string,
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  pipe(
    ShellCommand.make(command, url),
    ShellCommand.exitCode,
    Effect.mapError(() =>
      UsageError({message: 'Browser could not be opened.'}),
    ),
    Effect.map(() => void 0),
  );
