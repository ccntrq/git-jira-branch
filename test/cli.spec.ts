import {live} from '@effect/vitest';
import {Chunk, Effect, Option, pipe} from 'effect';
import type {Effect as EffectNs} from 'effect/Effect';

import {cliEffect} from '../src/cli';
import {
  getAssociatedBranches,
  gitCreateJiraBranch,
  switchBranch,
  ticketInfo,
  ticketInfoForCurrentBranch,
  ticketUrl,
  ticketUrlForCurrentBranch,
} from '../src/core';
import {openUrl} from '../src/url-opener';

import {type Mock, afterEach, describe, expect, vi} from 'vitest';
import {formatBranches} from '../src/branch-formatter';
import {formatIssue} from '../src/issue-formatter';
import {
  CreatedBranch,
  GitBranch,
  type JiraIssue,
  ResetBranch,
  SwitchedBranch,
  UsageError,
} from '../src/types';
import {dummyJiraIssue} from './dummies/dummyJiraIssue';
import {cliTestLayer} from './mock-implementations';
import {toEffectMock} from './util';

vi.mock('../src/core');
vi.mock('../src/url-opener');
vi.mock('../src/issue-formatter');
vi.mock('../src/branch-formatter');

const mockGitCreateJiraBranch = toEffectMock(
  gitCreateJiraBranch as unknown as Mock<
    Parameters<typeof gitCreateJiraBranch>,
    Effect.Effect<
      EffectNs.Success<ReturnType<typeof gitCreateJiraBranch>>,
      EffectNs.Error<ReturnType<typeof gitCreateJiraBranch>>
    >
  >,
);

const mockTicketUrl = toEffectMock(
  ticketUrl as unknown as Mock<
    Parameters<typeof ticketUrl>,
    Effect.Effect<
      EffectNs.Success<ReturnType<typeof ticketUrl>>,
      EffectNs.Error<ReturnType<typeof ticketUrl>>
    >
  >,
);

const mockTicketUrlForCurrentBranch = toEffectMock(
  ticketUrlForCurrentBranch as unknown as Mock<
    Parameters<typeof ticketUrlForCurrentBranch>,
    Effect.Effect<
      EffectNs.Success<ReturnType<typeof ticketUrlForCurrentBranch>>,
      EffectNs.Error<ReturnType<typeof ticketUrlForCurrentBranch>>
    >
  >,
);

const mockSwitchBranch = toEffectMock<
  typeof switchBranch
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
>(switchBranch as any);

const mockTicketInfoForCurrentBranch = toEffectMock<
  typeof ticketInfoForCurrentBranch
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
>(ticketInfoForCurrentBranch as any);

const mockTicketInfo = toEffectMock<
  typeof ticketInfo
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
>(ticketInfo as any);

const mockGetAssociatedBranches = toEffectMock<
  typeof getAssociatedBranches
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
>(getAssociatedBranches as any);

const mockOpenUrl = toEffectMock(
  openUrl as unknown as Mock<
    Parameters<typeof openUrl>,
    Effect.Effect<
      EffectNs.Success<ReturnType<typeof openUrl>>,
      EffectNs.Error<ReturnType<typeof openUrl>>
    >
  >,
);

const mockFormatIssue = formatIssue as unknown as Mock<[JiraIssue], string>;
const mockFormatBranches = formatBranches as unknown as Mock<
  [JiraIssue],
  string
>;

const mockLog = vi.spyOn(console, 'log').mockImplementation(() => undefined);

const withBaseArgs = (args: Array<string>): Effect.Effect<Array<string>> =>
  Effect.sync(() => ['node', 'git-jira-branch', ...args]);

describe('cli', () => {
  describe('main command', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should print version (--version)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['--version']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(mockLog.mock.calls[0]?.[0]).toMatch(/\d+\.\d+\.\d+/);
      }),
    );

    live('should print help (--help)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['--help']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );
  });

  describe('create subcommand', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should create branch with single argument', () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          CreatedBranch({branch: 'feat/FOOX-1234-description'}),
        );
        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['create', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
            ),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          'FOOX-1234',
          Option.none(),
          false,
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      }),
    );

    live('should inform about branch switch', () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          SwitchedBranch({branch: 'feat/FOOX-1234-description'}),
        );
        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['create', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
            ),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          'FOOX-1234',
          Option.none(),
          false,
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      }),
    );

    live('should handle basebranch argument (-b)', () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          CreatedBranch({branch: 'feat/FOOX-1234-description'}),
        );
        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['create', '-b', 'master', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
            ),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          'FOOX-1234',
          Option.some('master'),
          false,
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      }),
    );

    live('should handle reset option (-r)', () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          ResetBranch({branch: 'feat/FOOX-1234-description'}),
        );
        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['create', '-r', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
            ),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          'FOOX-1234',
          Option.none(),
          true,
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      }),
    );

    live('should report missing jirakey', () =>
      pipe(
        withBaseArgs(['create']),
        Effect.flatMap(cliEffect),
        Effect.match({
          onFailure: () => {
            expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
            expect(mockLog.mock.calls).toMatchSnapshot();
          },
          onSuccess: () => {
            expect.unreachable('Should have failed');
          },
        }),
        Effect.provide(cliTestLayer),
      ),
    );

    live('should print subcommand help (--help)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['create', '--help']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );

    live('should report unknwon argument', () =>
      pipe(
        withBaseArgs(['create', '111', 'unknown']),
        Effect.flatMap(cliEffect),
        Effect.match({
          onFailure: () => {
            expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
            expect(mockLog.mock.calls).toMatchSnapshot();
          },
          onSuccess: () => {
            expect.unreachable('Should have failed');
          },
        }),
        Effect.provide(cliTestLayer),
      ),
    );
  });

  describe('switch subcommand', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should switch branch', () =>
      Effect.gen(function* ($) {
        mockSwitchBranch.mockSuccessValue(
          SwitchedBranch({branch: 'feat/FOOX-1234-description'}),
        );
        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['switch', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
            ),
            cliTestLayer,
          ),
        );

        expect(mockSwitchBranch).toHaveBeenCalledWith('FOOX-1234');
        expect(mockLog.mock.calls).toMatchSnapshot();
      }),
    );

    live('should handle branch not found error', () =>
      Effect.gen(function* ($) {
        mockSwitchBranch.mockFailValue(
          UsageError({
            message: `No branch associated with Jira ticket 'FOOX-1234'`,
          }),
        );

        yield* $(
          Effect.provide(
            pipe(
              withBaseArgs(['switch', 'FOOX-1234']),
              Effect.flatMap(cliEffect),
              Effect.match({
                onFailure: () => expect(mockLog.mock.calls).toMatchSnapshot(),
                onSuccess: () => expect.unreachable('Should have failed'),
              }),
            ),
            cliTestLayer,
          ),
        );
      }),
    );

    live('should print subcommand help (--help)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['switch', '--help']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );
  });

  describe('open subcommand', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should open url for current branch without arguments', () =>
      Effect.gen(function* ($) {
        mockTicketUrlForCurrentBranch.mockSuccessValue(
          'https://gcjb.atlassian.net/browse/GCJB-1234',
        );
        mockOpenUrl.mockSuccessValue();
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['open']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockTicketUrlForCurrentBranch).toHaveBeenCalledWith();
        expect(mockLog.mock.calls).toMatchSnapshot();
        expect(mockOpenUrl.mock.calls).toMatchSnapshot();
      }),
    );

    live('should open url for given ticket', () =>
      Effect.gen(function* ($) {
        mockTicketUrl.mockSuccessValue(
          'https://gcjb.atlassian.net/browse/GCJB-1234',
        );
        mockOpenUrl.mockSuccessValue();
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['open', '1234']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockTicketUrl).toHaveBeenCalledWith('1234');
        expect(mockLog.mock.calls).toMatchSnapshot();
        expect(mockOpenUrl.mock.calls).toMatchSnapshot();
      }),
    );

    live('should print subcommand help (--help)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['open', '--help']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );
  });

  describe('info subcommand', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should render ticket info for current branch', () =>
      Effect.gen(function* ($) {
        mockTicketInfoForCurrentBranch.mockSuccessValue(dummyJiraIssue);
        mockFormatIssue.mockReturnValue('formatted issue');

        yield* $(
          withBaseArgs(['info']),
          Effect.flatMap(cliEffect),
          Effect.provide(cliTestLayer),
        );

        expect(mockTicketInfoForCurrentBranch).toHaveBeenCalledWith();
        expect(mockLog).toHaveBeenCalledWith('formatted issue');
        expect(mockFormatIssue).toHaveBeenCalledWith(dummyJiraIssue);
      }),
    );

    live('should render ticket info for given branch', () =>
      Effect.gen(function* ($) {
        mockTicketInfo.mockSuccessValue(dummyJiraIssue);
        mockFormatIssue.mockReturnValue('formatted issue');

        yield* $(
          withBaseArgs(['info', '1234']),
          Effect.flatMap(cliEffect),
          Effect.provide(cliTestLayer),
        );

        expect(mockTicketInfo).toHaveBeenCalledWith('1234');
        expect(mockLog).toHaveBeenCalledWith('formatted issue');
        expect(mockFormatIssue).toHaveBeenCalledWith(dummyJiraIssue);
      }),
    );

    live('should print subcommand help (--help)', () =>
      Effect.gen(function* ($) {
        yield* $(
          Effect.provide(
            pipe(withBaseArgs(['info', '--help']), Effect.flatMap(cliEffect)),
            cliTestLayer,
          ),
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );
  });

  describe('list subcommand', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should show associated branches', () =>
      Effect.gen(function* ($) {
        const associated = Chunk.fromIterable([
          GitBranch({
            name: 'feat/DUMMYAPP-123-dummy-issue-summary',
            isCurrent: true,
          }),
          GitBranch({
            name: 'feat/DUMMYAPP-124-another-issue-summary',
            isCurrent: false,
          }),
        ]);
        mockGetAssociatedBranches.mockSuccessValue(associated);
        mockFormatBranches.mockReturnValue('formatted branches');

        yield* $(
          withBaseArgs(['list']),
          Effect.flatMap(cliEffect),
          Effect.provide(cliTestLayer),
        );

        expect(mockFormatBranches).toHaveBeenCalledWith(associated);
        expect(mockLog).toHaveBeenCalledWith('formatted branches');
      }),
    );
  });
});
