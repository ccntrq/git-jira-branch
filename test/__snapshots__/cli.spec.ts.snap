// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > create subcommand > should create branch with single argument 1`] = `
[
  [
    "Successfully created branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > create subcommand > should handle basebranch argument (-b) 1`] = `
[
  [
    "Successfully created branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > create subcommand > should handle reset option (-r) 1`] = `
[
  [
    "Reset branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > create subcommand > should inform about branch switch 1`] = `
[
  [
    "Switched to already existing branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > create subcommand > should print subcommand help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ create [(-b, --base text)] [(-r, --reset)] <jira-key>

[0;1mDESCRIPTION[0m


Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.

[0;1mARGUMENTS[0m

[0;1m[0;30;1m<jira-key>[0;1m[0m

  A user-defined piece of text.

  The Jira ticket key to create a branch for (e.g. FOOX-1234)

[0;1mOPTIONS[0m

[0;1m(-b, --base text)[0m

  A user-defined piece of text.

  Base revision to create the new branch from (a branch name, tag or commit SHA)

  This setting is optional.

[0;1m(-r, --reset)[0m

  A true or false value.

  Reset the branch if it already exists

  This setting is optional.

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application

  This setting is optional.
"
`;

exports[`cli > create subcommand > should report missing jirakey 1`] = `
[
  [
    "Missing argument <jira-key>
",
  ],
]
`;

exports[`cli > create subcommand > should report unknwon argument 1`] = `
[
  [
    "Received unknown argument: 'unknown'
",
  ],
]
`;

exports[`cli > info subcommand > should print subcommand help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ info [<jira-key>]

[0;1mDESCRIPTION[0m


Displays information for the given Jira ticket on your terminal. If no ticket is
provided, it presents information for the Jira ticket associated with the
current branch.

[0;1mARGUMENTS[0m

[0;1m[0;30;1m<jira-key>[0;1m[0m

  A user-defined piece of text.

  The Jira key for the ticket to get information for (e.g. FOOX-1234)

  This setting is optional.

[0;1mOPTIONS[0m

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application

  This setting is optional.
"
`;

exports[`cli > main command > should print help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ git-jira-branch

[0;1mOPTIONS[0m

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application

  This setting is optional.

[0;1mCOMMANDS[0m

  - create [(-b, --base text)] [(-r, --reset)] <jira-key>  
Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.

  - switch <jira-key>                                      
Switches to an already existing branch that is associated with the given Jira
ticket.

  - open [<jira-key>]                                      
Opens the given Jira ticket in your default browser. If no ticket is given the
jira ticket for the current branch is opened.

  - info [<jira-key>]                                      
Displays information for the given Jira ticket on your terminal. If no ticket is
provided, it presents information for the Jira ticket associated with the
current branch.

  - list                                                   
Lists all branches that appear to be associated with a Jira ticket.
"
`;

exports[`cli > open subcommand > should open url for current branch without arguments 1`] = `
[
  [
    "Opening ticket url 'https://gcjb.atlassian.net/browse/GCJB-1234' in your default browser...",
  ],
]
`;

exports[`cli > open subcommand > should open url for current branch without arguments 2`] = `
[
  [
    "https://gcjb.atlassian.net/browse/GCJB-1234",
  ],
]
`;

exports[`cli > open subcommand > should open url for given ticket 1`] = `
[
  [
    "Opening ticket url 'https://gcjb.atlassian.net/browse/GCJB-1234' in your default browser...",
  ],
]
`;

exports[`cli > open subcommand > should open url for given ticket 2`] = `
[
  [
    "https://gcjb.atlassian.net/browse/GCJB-1234",
  ],
]
`;

exports[`cli > open subcommand > should print subcommand help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ open [<jira-key>]

[0;1mDESCRIPTION[0m


Opens the given Jira ticket in your default browser. If no ticket is given the
jira ticket for the current branch is opened.

[0;1mARGUMENTS[0m

[0;1m[0;30;1m<jira-key>[0;1m[0m

  A user-defined piece of text.

  The Jira key for the ticket to open (e.g. FOOX-1234)

  This setting is optional.

[0;1mOPTIONS[0m

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application

  This setting is optional.
"
`;

exports[`cli > switch subcommand > should handle branch not found error 1`] = `
[
  [
    "[0;31mNo branch associated with Jira ticket 'FOOX-1234'[0m
",
  ],
]
`;

exports[`cli > switch subcommand > should print subcommand help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ switch <jira-key>

[0;1mDESCRIPTION[0m


Switches to an already existing branch that is associated with the given Jira
ticket.

[0;1mARGUMENTS[0m

[0;1m[0;30;1m<jira-key>[0;1m[0m

  A user-defined piece of text.

  The Jira ticket key associated with the branch to switch to (e.g. FOOX-1234)

[0;1mOPTIONS[0m

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application

  This setting is optional.
"
`;

exports[`cli > switch subcommand > should switch branch 1`] = `
[
  [
    "Switched to already existing branch: 'feat/FOOX-1234-description'",
  ],
]
`;
